### PR TITLE
hra-download-state behavioral

### PR DESCRIPTION
**Pull request by sanket1211996**

### CODE REVIEW

File: libs/state/src/lib/download-file/download-file.state.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * Actions download file in specified format
+   * @param ctx
+   * @param action
+   */
+  @Action(DownloadFile)
+  async downloadFile(ctx: StateContext<DownloadStateModel>, action: DownloadFile) {
+    const fileName = fileUrl.substring(fileUrl.lastIndexOf('/') + 1, fileUrl.lastIndexOf('.'));
+
```
Comment (axdanbol) [inline libs/state/src/lib/download-file/download-file.state.ts:48]: If url is available then download directly. Otherwise if format is supported convert

**Review by axdanbol — CHANGES_REQUESTED** [review]

**Comment by sanket1211996** [comment]
@bherr2  @axdanbol Code is ready for review there is some issue with checks. Could you please check.

File: libs/state/src/lib/download/download.model.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+   */
+  format: FileFormat;
+
+  /**
+   * label display in UI
+   */
+  label: string;
+}
+
+export type DownloadModel = DownloadFormat[];
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.model.ts:26]: Add doc comment

File: libs/state/src/lib/download/index.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export * as DownloadAction from './download.action';
+export { DownloadState } from './download.state';
+export { FileFormat, DownloadFormat } from './download.model';
```
Comment (axdanbol) [inline libs/state/src/lib/download/index.ts:3]: Export selectors

File: libs/state/src/lib/download/download.state.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    },
+  ],
+})
+export class DownloadState {
+  /**
+   * Download converted formatted file to browser
+   * @param blob
+   * @param fileName
+   */
+  blobDownload(blob: Blob, fileName: string) {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:32]: Rename to `downloadData`
Should also be private

File: libs/state/src/lib/download/download.state.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  inferFilename(fileUrl: string) {
+    return fileUrl.substring(fileUrl.lastIndexOf('/') + 1, fileUrl.lastIndexOf('.'));
+  }
+
+  /**
+   * Infers file extension from file url
+   * @param fileUrl
+   * @returns
+   */
+  inferFileExtension(fileUrl: string) {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:57]: Don't need to guess extension. You get that from the format

File: libs/state/src/lib/download/download.state.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    return fileUrl.substring(fileUrl.lastIndexOf('.') + 1);
+  }
+
+  /**
+   * Downloads and save -  method is used to direct fetch file
+   * if available on url without conversion
+   * @param fileUrl
+   * @param fileName
+   */
+  async downloadAndSave(fileUrl: string, fileName: string) {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:67]: Rename to `downloadRemoteData`

File: libs/state/src/lib/download/download.state.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    anchor.remove();
+    window.URL.revokeObjectURL(url);
+  }
+
+  /**
+   * Infers filename from file url
+   * @param fileUrl
+   * @returns
+   */
+  inferFilename(fileUrl: string) {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:48]: Rename to `guessFilename`
Also it doesn't work as intended. Try the following examples:
`abc`
`abc/def`
`abc/def/`

File: libs/state/src/lib/download/download.state.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+   * Creates canvas of svg from HTMLImageElement
+   * @param surl
+   * @returns canvas
+   */
+  async createCanvas(image: HTMLImageElement): Promise<HTMLCanvasElement> {
+    const canvas = document.createElement('canvas');
+    const ctx = canvas.getContext('2d');
+    canvas.width = image.width;
+    canvas.height = image.height;
+    ctx?.drawImage(image, 0, 0);
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:122]: Should fail if we can't draw it (i.e. no context)

File: libs/state/src/lib/download/download.state.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    await image.decode();
+    return image;
+  }
+
+  /**
+   * Creates canvas of svg from HTMLImageElement
+   * @param surl
+   * @returns canvas
+   */
+  async createCanvas(image: HTMLImageElement): Promise<HTMLCanvasElement> {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:117]: Rename to `convertImageToCanvas`

**Review by axdanbol — CHANGES_REQUESTED** [review]

File: libs/state/src/lib/download/download.action.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+export class AddEntry extends Action('Add') {
+  constructor(readonly id: DownloadFormatId, readonly entry: DownloadEntry) {
+    super();
+  }
+}
+
+/**
+ * Clear entry from download state
+ */
+export class ClearEntry extends Action('Clear') {
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.action.ts:27]: I rather have an action that clears all entries (i.e. sets entries to the empty object {}) than having to clear each one by one

File: libs/state/src/lib/download/download.selectors.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import { Selector } from '@ngxs/store';
+import { DownloadFormat, DownloadModel } from './download.model';
+
+/**
+ * Available format selectors
+ */
+export class DownloadSelectors {
+  /**
+   * Selectors available format for download selectors
+   * @param state
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.selectors.ts:10]: Add docs to param and return tags

File: libs/state/src/lib/download/download.state.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+ */
+@State<DownloadModel>({
+  name: 'download',
+  defaults: {
+    formats: {},
+    entries: {},
+  },
+})
+export class DownloadState implements NgxsOnInit {
+  private readonly http = inject(HttpClient);
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:24]: Code doc

File: libs/state/src/lib/download/download.state.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    const { entries } = ctx.getState();
+    const entry = entries[format];
+    let filename = `download${format}`;
+    switch (entry?.type) {
+      case 'url':
+        filename = this.guessFilename(ctx, format, entry.url);
+        return this.downloadRemoteData(entry.url).pipe(tap((data) => this.downloadData(data, filename)));
+
+      case 'data':
+        this.downloadData(new Blob([entry.data]), filename);
```
Comment (axdanbol) [inline libs/state/src/lib/download/download.state.ts:92]: Use guessFilename with an empty url '' for the filename here

**Review by axdanbol — CHANGES_REQUESTED** [review]

**Review by axdanbol — APPROVED** [review]

**Pull request merged**